# **HANDS-ON 3: Criando os Primeiros Neurônios**

Como vimos anteriormente, as redes neurais artificiais foram criadas com base em um cérebro orgânico. Apesar dos dois sistemas serem essencialmente diferentes, podemos realizar algumas comparações: Ambos possuem um elemento chamado de neurônio - palavra que deriva do grego e significa "nervo" - e que servem para a propagação/modificação de sinais.

![](../Imagens/Unidade%201/modelo_neuronio.png)

Link da imagem: [Clique Aqui](https://en.wikipedia.org/wiki/Artificial_neuron#/media/File:Neuron3.svg)

Um único neurônio não é capaz de realizar grandes operações, contudo, quando em operação com diveros outros neurônios, o sistema adquire a robustez de operacionalizar tarefas complexas. Na imagem abaixo, temos um exemplo de rede neuronal.

![](../Imagens/Unidade%201/exemplo_rede_neural.png)



No decorrer do repositório, iremos utilizar a teoria Hebbeliana, isto é, um aumento na eficácia sináptica surge da estimulação repetida e persistente de uma célula pré-sináptica de uma célula pós-sináptica. O princípio de Hebb pode ser descrito como um método para determinar como alterar os pesos entre os neurônios modelo. O peso entre dois neurônios aumenta se os dois neurônios forem ativados simultaneamente e reduz se eles forem ativados separadamente. Os nós que tendem a ser positivos ou negativos ao mesmo tempo têm pesos positivos fortes, enquanto aqueles que tendem a ser opostos têm pesos negativos fortes.

![](../Imagens/Unidade%201/exemplo_neuronio.png)

Iremos utilizar a imagem acima para codificar o primeiro neurônio. Note podemos expressar a saida como sendo uma função - que o neurônio realiza - sobre cada entrada multiplicada pelo seu respectivo peso. Se denotarmos as entradas por $x[n]$, os pesos por $w[n]$ e $b$ uma constante, então:

$$Output = x[1] * w[1] + x[2] * w[2] + x[3] * w[3] + \cdots + x[n] * w[n] + b$$

### **Exemplo 01 - O primeiro neurônio:**

In [28]:
input = [0, 1, 2]          # Entradas (Inputs) que serão feitas
weight = [0.9, -1.3, 1.5]  # Pesos (Weights) que estão associados a cada entrada
bias = 3.33                # Bias (Viés/Constante)

# Saida
output = (input[0]*weight[0] +
          input[1]*weight[1] +
          input[2]*weight[2] + bias)

# Printa a saida do neurônio
print(f'A saida de seu neurônio é: {output}')

A saida de seu neurônio é: 5.03


Agora que entendemos como podemos programar um neurônio, iremos verificar como podemos expandir esse conceito e construir mais de um nerurônio. Neste segundo exemplo, iremos programar 4 neurônios.

### **Exemplo 02 - Quatro neurônios**

In [29]:
input = [0, 1, 2]           # Entradas (Inputs) que serão feitas

weight_0 = [0.9, -1.3, 1.5] # Pesos (Weights) do primeiro neurônio
weight_1 = [0.8, -1.4, 1.4] # Pesos (Weights) do segundo neurônio
weight_2 = [0.7, -1.5, 1.6] # Pesos (Weights) do terceiro neurônio
weight_3 = [0.6, -1.6, 1.3] # Pesos (Weights) do quarto neurônio

bias_0 = 3.33               # Bias (Viés/Constante) do primeiro neurônio
bias_1 = 4.44               # Bias (Viés/Constante) do segundo neurônio
bias_2 = 2.22               # Bias (Viés/Constante) do terceiro neurônio
bias_3 = 1.11               # Bias (Viés/Constante) do quarto neurônio

# Primeiro Neurônio
output_0 = (input[0]*weight_0[0] +
            input[1]*weight_0[1] +
            input[2]*weight_0[2] + bias_0)

# Segundo Neurônio
output_1 = (input[0]*weight_1[0] +
            input[1]*weight_1[1] +
            input[2]*weight_1[2] + bias_1)

# Terceiro Neurônio
output_2 = (input[0]*weight_2[0] +
            input[1]*weight_2[1] +
            input[2]*weight_2[2] + bias_2)

# Quarto Neurônio
output_3 = (input[0]*weight_3[0] +
            input[1]*weight_3[1] +
            input[2]*weight_3[2] + bias_3)

print(f'A saida do primeiro neurônio é: {output_0}')
print(f'A saida do segundo neurônio é: {output_1}')
print(f'A saida do terceiro neurônio é: {output_2}')
print(f'A saida do quarto neurônio é: {output_3}')

A saida do primeiro neurônio é: 5.03
A saida do segundo neurônio é: 5.84
A saida do terceiro neurônio é: 3.9200000000000004
A saida do quarto neurônio é: 2.1100000000000003


### **Exemplo 3 - Quarenta Neurônios**

Para programarmos 40 neurônios, iremos mudar a abordagem e automzatizar os resultados com a biblioteca **Random**.

In [30]:
# Importando bibliotecas
from random import random as rnd

inputs = [rnd() for i in range(10)]                          # Gerando 10 números aleatórios de inputs 
weights = {i: [rnd() for j in range(10)] for i in range(40)} # Criando 40 neurônios com pesos aleatórios
bias = [rnd() for i in range(10)]                            # Criando 10 viés/constantes com valores aleatórios

saidas = []                                                  # Lista de saídas
for neuron_wright, neuron_bias in zip(weights.values(), bias):
    saidas.append(sum([i*j for i, j in zip(inputs, neuron_wright)]) + neuron_bias)

print(saidas)

[3.664904334457117, 2.275832448488788, 3.7318904699380164, 2.631760979962661, 2.4652874687221473, 2.6010362017203836, 3.042924939290355, 2.7624137141142002, 2.903883777509586, 3.4592468522287536]


Como podemos ver, realizar manualmente a programação dos naurônios é custoso e, devido a esse pressuposto, iremos utilizar a biblioteca **NumPy**. NumPy é uma biblioteca que suporta o processamento de grandes dados, multi-dimensionais arranjos e matrizes, juntamente com uma grande coleção de funções matemáticas de alto nível para operar sobre esses entes. 

Além de poupar tempo na programação, NumPy é uma biblioteca que foi desenvolvida para otimizar atividades em Python.
* O pacote NumPy divide uma tarefa em vários fragmentos e processa todos os fragmentos paralelamente.
* O pacote NumPy integra códigos C, C++ e Fortran em Python. Essas linguagens de programação têm muito pouco tempo de execução em comparação com o Python.

### **Exemplo 4 - Novamente, um neurônio, mas com numpy**

In [31]:
# Importando bibliotecas
import numpy as np

input = [0, 1, 2]          # Entradas (Inputs) que serão feitas
weight = [0.9, -1.3, 1.5]  # Pesos (Weights) que estão associados a cada entrada
bias = 3.33                # Bias (Viés/Constante)

# Saida
output = np.dot(input, weight) + bias

# Printa a saida do neurônio
print(f'A saida de seu neurônio é: {output}')

A saida de seu neurônio é: 5.03


### **Exemplo 5 - Novamente, 40 neurônios, mas com numpy**

In [32]:
import numpy as np

inputs = [rnd() for i in range(10)]                          # Gerando 10 números aleatórios de inputs 
weights = {i: [rnd() for j in range(10)] for i in range(40)} # Criando 40 neurônios com pesos aleatórios
bias = [rnd() for i in range(10)]                            # Criando 10 viés/constantes com valores aleatórios

saidas = []
for neuron_wright, neuron_bias in zip(weights.values(), bias):
    saidas.append(np.dot(inputs, neuron_wright) + neuron_bias)

print(saidas)

[1.18322023870852, 2.190592024165915, 1.700489958183052, 1.8532850120226092, 1.7850856815965495, 1.8519331633835754, 2.519266934817108, 2.3186750158933678, 1.7917776347169412, 1.874607443326167]


**Relatório 03:** 

1 - **Sem utilizar nenhuma biblioteca, faça um código em Python que modela apenas um neurônio. Esse nerônio deve ter como inputs**:

$$inputs = [1.2, 0.45, -1.5, 1.9, 0.6]$$

*e de ve ter como pesos*:

$$weights = [1.5, 2.4, -2, 0.45, 2.8]$$

2 - **Sem utilizar nenhuma biblioteca, faça um código em Python que modela apenas três neurônios. Esses nerônios deve ter como inputs**:

$$inputs = [1.1, 2.2, 3.3, 4.4]$$

*e deve ter como pesos*:

$$weights = [[0.2, 0.8, -0.5, 1],
[0.5, -0.91, 0.26, -0.5],
[-0.26, -0.27, 0.17, 0.87]]$$

3 - **Utlizando as bibliotecas random e numpy, faça um código em Python que modela vinte mil neurônios.**:

4 - **Compare o tempo de execução do script com uso da biblioteca numpy com o script sem a biblioteca. OBS: Se necessário, utilize a biblioteca "time"**

# **Referência Bibliográfica**

1 - Teoria Hebbeliana. Disponível em: <https://en.wikipedia.org/wiki/Hebbian_theory>. Acessado em 10 de maio de 2022

2 - Bilioteca Numpy. Disponível em: <https://pt.wikipedia.org/wiki/NumPy>. Acessado em 10 de maio de 2022.

3 - Why is Numpy faster in Python?. Disponível em: <https://www.geeksforgeeks.org/why-numpy-is-faster-in-python/#:~:text=NumPy%20Arrays%20are%20faster%20than,in%20non%2Dcontiguous%20memory%20locations.>. Acessado em 10 de maio de 2022.